### PR TITLE
wence/fix/interpolate max

### PR DESCRIPTION
**Pull request by wence-**
Addresses #1768.

### CODE REVIEW

**Comment by wence-** [comment]
Needs OP2/PyOP2#590.

**Comment by sv2518** [comment]
I think the initialisation is not quite right. We also need to initialise empty functions. With the following the test passes.
```
diff --git a/firedrake/interpolation.py b/firedrake/interpolation.py
index 97688a71..1bd73747 100644
--- a/firedrake/interpolation.py
+++ b/firedrake/interpolation.py
@@ -131,6 +131,8 @@ def make_interpolator(expr, V, subset, access):
             V = f.function_space()
         else:
             f = firedrake.Function(V)
+        if ((isinstance(V, firedrake.Function) and not hasattr(f._data.dat, "_numpy_data"))
+           or not isinstance(V, firedrake.Function)):
             if access in {firedrake.MIN, firedrake.MAX}:
                 finfo = numpy.finfo(f.dat.dtype)
                 if access == firedrake.MIN:
```

**Comment by wence-** [comment]
> I think the initialisation is not quite right. We also need to initialise empty functions.

The behaviour was correct, but the test was wrong. The model is that if you interpolate into a function then that functions values should take part in the min/maxing. If you interpolate into a space, then they shouldn't. I have fixed the test.

**Comment by wence-** [comment]
> The model is that if you interpolate into a function then that functions values should take part in the min/maxing.

Added this to the docstring.

**Comment by sv2518** [comment]
Ah, ok. Thanks!

File: firedrake/mg/utils.py [hunk 1]
```diff
@@ -140,6 +140,8 @@ def physical_node_locations(V):
         return cache[key]
     except KeyError:
         Vc = firedrake.FunctionSpace(mesh, ufl.VectorElement(element))
+        locations = firedrake.Function(Vc)
```
Comment (pefarrell) [inline firedrake/mg/utils.py:143]: Maybe a dumb question: doesn't this create a variable, then immediately overwrite it?

Comment (wence-) [reply firedrake/mg/utils.py:143]: Oh yeah!

**Comment by pefarrell** [comment]
It would be nice to have a test where we take a periodic mesh (with DG coordinates), interpolate using `MAX` to a CG space, and check that the coordinates make sense (the choice of coordinates on the seam has been consistently taken from one side).

**Comment by wence-** [comment]
> It would be nice to have a test where we take a periodic mesh (with DG coordinates), interpolate using MAX to a CG space, and check that the coordinates make sense (the choice of coordinates on the seam has been consistently taken from one side).

Did that, will merge.

File: firedrake/mg/utils.py (hunk 1 above)
Comment (wence-) [reply firedrake/mg/utils.py]: Fixed.

**Pull request closed without merging**